File: src/main/java/com/xq/tabletest/udftest/AggregateFunctionTest.java
package com.xq.tabletest.udftest;

import com.xq.apitest.pojo.SensorReading;
import com.xq.tabletest.TableApiTest;
import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.TimestampAssigner;
import org.apache.flink.api.common.eventtime.TimestampAssignerSupplier;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.table.api.DataTypes;
import org.apache.flink.table.api.EnvironmentSettings;
import org.apache.flink.table.api.Expressions;
import org.apache.flink.table.api.Schema;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.types.Row;

import java.net.URL;
import java.time.Duration;

import static org.apache.flink.table.api.Expressions.$;
import static org.apache.flink.table.api.Expressions.call;
import static org.apache.flink.table.api.Expressions.sourceWatermark;

public class AggregateFunctionTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        EnvironmentSettings settings = EnvironmentSettings.newInstance().inStreamingMode().build();
        StreamTableEnvironment tabEnv = StreamTableEnvironment.create(env, settings);

        URL resource = TableApiTest.class.getResource("/sensor.txt");
        String filePath = resource.getPath();
        DataStreamSource<String> inputStream = env.readTextFile(filePath);

        WatermarkStrategy<SensorReading> watermarkStrategy = WatermarkStrategy.<SensorReading>forBoundedOutOfOrderness(Duration.ofMillis(1000L))
                .withTimestampAssigner((SerializableTimestampAssigner<SensorReading>) (element, recordTimestamp) -> element.getTimestamp());

        // 先转换成样例类类型（简单转换操作）
        SingleOutputStreamOperator<SensorReading> dataStream = inputStream.map((MapFunction<String, SensorReading>) value -> {
            String[] split = value.split(",");
            return new SensorReading(split[0].trim(), Long.parseLong(split[1].trim()), Double.parseDouble(split[2].trim()));
        })
                .assignTimestampsAndWatermarks(watermarkStrategy);

        Schema schema = Schema.newBuilder()
                .column("id", DataTypes.STRING())
                .column("temperature", DataTypes.DOUBLE())
                .columnByExpression("temp1", "temperature")
                .columnByExpression("temp", "cast(temperature as DOUBLE)")
//                .column("timestamp", DataTypes.BIGINT())
//                .columnByExpression("ts", Expressions.callSql("TO_TIMESTAMP_LTZ(`timestamp`, 0)"))
                .columnByExpression("ts", "TO_TIMESTAMP_LTZ(`timestamp`, 0)")//事件时间，类是字段与sql关键字冲突，要加上`号
                .columnByExpression("ps", "PROCTIME()")//操作时间
                .watermark("ts", "SOURCE_WATERMARK()")
//                .watermark("ts", sourceWatermark())
                .build();
        Table sensorTable = tabEnv.fromDataStream(dataStream, schema);
//        Table sensorTable = tabEnv.fromDataStream(dataStream).as("id","ts","temperature");

//        tabEnv.registerFunction("avgTmp",new AvgTemp());
        tabEnv.createTemporarySystemFunction("avgTmp",new AvgTemp());
        Table select = sensorTable.groupBy($("id"))
//                .aggregate("avgTmp(temperature) as avgTemp")
                .aggregate(call("avgTmp", $("temp")).as("avgTemp"))
                .select($("id"),$("avgTemp"));
//        tabEnv.toRetractStream(select, Row.class).print("tst");
        tabEnv.toChangelogStream(select).print("tst");

        env.execute("agg func");
    }
}

File: src/main/scala/com/atguigu/apitest/tabletest/udftest/AggregateFunctionTest.scala
package com.atguigu.apitest.tabletest.udftest

import org.apache.flink.api.scala._
import com.atguigu.apitest.SensorReading
import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.table.api._
import org.apache.flink.table.api.bridge.scala._
import org.apache.flink.table.functions.AggregateFunction
import org.apache.flink.types.Row

import java.time.Duration

/**
  * Copyright (c) 2018-2028   All Rights Reserved
  *
  * Project: FlinkTutorial
  * Package: com.atguigu.apitest.tabletest.udftest
  * Version: 1.0
  *
  * Created by hr on 2020/6/1 15:38
  */
object AggregateFunctionTest {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
//    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    // 创建表执行环境
    val tableEnv = StreamTableEnvironment.create(env)

    val resource = getClass.getClassLoader.getResource("sensor.txt")
    val filePath = resource.getPath
    val inputStream: DataStream[String] = env.readTextFile(filePath)
    //    val inputStream: DataStream[String] = env.socketTextStream("localhost", 7777)

    val watermarkStrategy = WatermarkStrategy.forBoundedOutOfOrderness[SensorReading](Duration.ofMillis(1000L))
      .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
        override def extractTimestamp(element: SensorReading, recordTimestamp: Long): Long = element.timestamp
      })

    // map成样例类类型
    val dataStream: DataStream[SensorReading] = inputStream
      .map(data => {
        val dataArray = data.split(",")
        SensorReading(dataArray(0), dataArray(1).toLong, dataArray(2).toDouble)
      })
      .assignTimestampsAndWatermarks(watermarkStrategy)

    val schema = Schema.newBuilder()
      .columnByExpression("ts", "TO_TIMESTAMP_LTZ(`timestamp`, 0)")
      .column("id", "STRING")
      .column("temperature", "DOUBLE")
//      .columnByExpression("temp1", "temperature")
//      .columnByExpression("temp", "cast(temperature as DOUBLE)")
       //.column("timestamp", DataTypes.BIGINT())
      //.columnByExpression("ts", Expressions.callSql("TO_TIMESTAMP_LTZ(`timestamp`, 0)"))
//      .columnByExpression("ps", "PROCTIME()") //事件时间，类是字段与sql关键字冲突，要加上`号
     //操作时间
      .watermark("ts", "SOURCE_WATERMARK()")
      //.watermark("ts", sourceWatermark())
      .build()
    val sensorTable = tableEnv.fromDataStream(dataStream, schema)
    // 将流转换成表，直接定义时间字段
//    val sensorTable: Table = tableEnv.fromDataStream(dataStream, schema)

    // 先创建一个聚合函数的实例
    val avgTemp = new AvgTemp()

    // Table API 调用
    val resultTable = sensorTable
      .groupBy('id)
      .aggregate( avgTemp('temperature) as 'avgTemp )
      .select('id, 'avgTemp)

    resultTable.toRetractStream[Row].print("result")

    env.execute("agg udf test job")
  }

  // 专门定义一个聚合函数的状态类，用于保存聚合状态（sum，count）
  class AvgTempAcc {
    var sum: Double = 0.0
    var count: Int = 0
  }
  // 自定义一个聚合函数
  class AvgTemp extends AggregateFunction[Double, AvgTempAcc]{

    override def getValue(accumulator: AvgTempAcc): Double = accumulator.sum/accumulator.count

    override def createAccumulator(): AvgTempAcc = new AvgTempAcc()

    def accumulate(acc: AvgTempAcc, temp: Double): Unit ={
      acc.sum += temp
      acc.count += 1
    }
  }
}
